### PR TITLE
[ENH] Add PyOD adapter for anomaly detection

### PR DESCRIPTION
**Pull request by CodeLionX**
#### Reference Issues/PRs

Requires #1605

#### What does this implement/fix? Explain your changes.

Adds an adapter class for PyOD anomaly detection models. With this adapter, we cover the following algorithms from the evaluation:

- Sub-LOF `PyODAdapter(LOF(...), window_size=ws, stride=1)` with `ws > 1`
- LOF `PyODAdapter(LOF(...), window_size=1, stride=1)`
- Sub-IF ...
- iForest
- KNN
- HBOS
- CBLOF
- COPOD
- COF

And [many more algorithms](https://github.com/yzhao062/pyod/tree/development?tab=readme-ov-file#implemented-algorithms) that were not part of the evaluation.

The tests are still experimental because I needed to mock the PyOD models. Feedback is welcome.

#### Does your contribution introduce a new dependency? If yes, which one?

- [PyOD](https://pyod.readthedocs.io/en/latest/) as a soft-dependency

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [x] I've added the estimator to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [x] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/anomaly_detection/_pyodadapter.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+        Returns
+        -------
+        params : dict or list of dict, default={}
+            Parameters to create testing instances of the class.
+            Each dict are parameters to construct an "interesting" test instance, i.e.,
+            `MyClass(**params)` or `MyClass(**params[i])` creates a valid test instance.
+            `create_test_instance` uses the first (or only) dictionary in `params`.
+        """
+        return {"pyod_model": _PyODMock(), "window_size": 5, "stride": 1}
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/_pyodadapter.py:131]: Think an actual model should be fine here if you want to use it, just import it in the functions.

File: aeon/anomaly_detection/_pyodadapter.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    stride : int, default=1
+        Stride of the sliding window.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from pyod.models.lof import LOF
+    >>> from aeon.anomaly_detection import PyODAdapter
+    >>> X = np.random.default_rng(42).random((10, 2), dtype=np.float_)
+    >>> detector = PyODAdapter(LOF(), window_size=2)
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/_pyodadapter.py:51]: you'll want to skip the lines which require optional dependencies using `# doctest: +SKIP`

**Review by MatthewMiddlehurst — COMMENTED** [review]
looks good for the most part.

File: aeon/anomaly_detection/tests/test_pyod_adapter.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+import numpy as np
+
+from aeon.anomaly_detection import PyODAdapter
+from aeon.anomaly_detection._pyodadapter import _PyODMock
+from aeon.testing.utils.data_gen import make_series
+
+
+def test_pyod_adapter_default():
+    """Test PyODAdapter."""
```
Comment (MatthewMiddlehurst) [inline aeon/anomaly_detection/tests/test_pyod_adapter.py:13]: You will want to skip these conditionally with a decorator, these are the ones used for tsfresh.

```
@pytest.mark.skipif(
    not _check_soft_dependencies("tsfresh", severity="none"),
    reason="skip test if required soft dependency tsfresh not available",
)
```

Can just swap out the dependency with pyod probably.

**Review by MatthewMiddlehurst — COMMENTED** [review]
There is a conflict currently, but LGTM after.

**Comment by MatthewMiddlehurst** [comment]
I'll jump in and see what I can do. Annoying stuff like this is why i'm trying to revamp it all.

**Review by MatthewMiddlehurst — APPROVED** [review]
soft dependency issues seem to be resolved. Notebook issue can be fixes here or in a separate PR.

**Comment by CodeLionX** [comment]
> soft dependency issues seem to be resolved. Notebook issue can be fixes here or in a separate PR.

should be fixed with #1695

**Pull request merged**